### PR TITLE
Breaking: Ajv to v8.5.0, added ajv-draft-04 (fixes #13888)

### LINKED ISSUE #13888: Upgrade to Ajv version 8
**The version of ESLint you are using.**

7.14

**The problem you want to solve.**

Migrate eslint to Ajv version 7 at the time it is released in December as the main version, to ensure that the users migrating to Ajv v7 and using eslint (that depends on Ajv version 6) do not have any version conflicts.

**Your take on the correct solution to problem.**

Upgrade in eslint: 
- Ajv and other related dependencies versions (e.g. ajv-keywords)
- any Ajv API/options changes
- JSON Schema to draft-07 - Ajv v7 removed support for draft-04 (which is 4 versions old), the change logs are here: https://github.com/ajv-validator/ajv/releases

My concern that any plugins that rely on JSON Schema draft-04 specific features may need to be upgraded too, but nothing has been reported so far; although not too many users started using Ajv v7 beta, it is growing judging by ajv-formats downloads that was split from Ajv in v7: https://www.npmjs.com/package/ajv-formats

Either way, it most likely should be a major version change in eslint too to avoid any issues for the users - do you agree?

I am planning a release candidate version 7.0.0-rc.0 some time next week, and the main version release within a couple of weeks this year; I think I should  make a PR to eslint once 7.0.0-rc.0 is released.

Let me know if you have any concerns / comments / questions

**Are you willing to submit a pull request to implement this change?**

Yes

### PR DESCRIPTION
**Pull request by epoberezkin**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/master/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[x] Other, please explain: Update Ajv version to v7

#### What changes did you make? (Give an overview)

- Update Ajv to version v7.0.0-rc.0
- Update schemas:
  - remove ignored `additionalItems` keywords (it is possible though that the intention was to have only one item allowed in these schemas, in which case schemas have to be changed from `items: {...}` to `items: [{...}], additionalItems: false`
  - add missing `type` keywords (in the second commit).
- Remove `strictDefaults` option ([strict mode](https://github.com/ajv-validator/ajv/blob/master/docs/strict-mode.md) is on by default now) and other deprecated option
- Use JSON Schema draft-07 (although it was not explicitly specified that schemas are draft-04)
- Fix failing tests (all updates are because of changes in how dataPath is shown)

#### Is there anything you'd like reviewers to focus on?

- Review updated schemas whether they are correct.
- Please note that Ajv v7 no longer supports JSON Schema draft-04, but all schemas I have seen are draft-07 compatible, so no changes were necessary. As I wrote in #13888, it may have effect on plugins maybe? It is probably still worth trying to release as a minor version upgrade (once Ajv v7 becomes the main version in 1-2 weeks) and monitor if there are some issues while resolving them.

cc @nzakas

### CODE REVIEW

**Comment by mdjermanovic** [comment]
@epoberezkin thanks for the PR! I marked this PR as a draft because of the rc version, and as "breaking" until we figure out whether or not it's a breaking change.

File: lib/rules/no-restricted-modules.js
Comment (mdjermanovic) [inline lib/rules/no-restricted-modules.js:68]: This should have been a 1-element tuple, but since it's a deprecated rule it's fine to not change anything.

**Comment by mdjermanovic** [comment]
> * add missing `type` keywords (in the second commit)

How was this working at all, does Ajv in non-strict mode assume it's an "object"/"array" type if there is "properties"/"items"?

> * [strict mode](https://github.com/ajv-validator/ajv/blob/master/docs/strict-mode.md) is on by default now

Should we then turn it off explicitly, to retain the previous behavior (and thus maybe avoid breaking plugins)?

> * Please note that Ajv v7 no longer supports JSON Schema draft-04, but all schemas I have seen are draft-07 compatible, so no changes were necessary. As I wrote in #13888, it may have effect on plugins maybe?

Is there an overview of the differences between versions, so we could maybe try to estimate the impact on plugins? In particular, what was valid by draft-04, but would be invalid or ignored or maybe have different semantics with draft-07.

File: lib/rules/no-restricted-modules.js
Comment (epoberezkin) [reply lib/rules/no-restricted-modules.js:68]: there were couple of other places - probably worth making them all 1-tuples if that was the intention?

**Comment by epoberezkin** [comment]
>>add missing type keywords (in the second commit)

>How was this working at all, does Ajv in non-strict mode assume it's an "object"/"array" type if there is "properties"/"items"?

No, it allows ANY data type, "property"/"items" only applies any restriction IF it is "object"/"array" - this is how JSON Schema specifies it. So if the data were an object/array it would validate it as expected, otherwise it would just allow it (incorrectly). It is a common mistake, that's why I made strict mode default.

>> strict mode is on by default now

> Should we then turn it off explicitly, to retain the previous behavior (and thus maybe avoid breaking plugins)?

We could, but:

- there may be other schema mistakes that would remain unnoticed
- the main thing is that disabling strict mode won't guarantee that this is not a breaking change, because if plugins rely on eslint code validating their config schemas (am I correct - is this how it is?), they may use some draft-04 specific things that won't be supported any more:
  - "id" (rather than $id), although it is probably not very likely - most schemas don't seem to rely on schema "id"
  - old-style "exclusiveMaximum"/"exclusiveMinimum" keywords

I believe we should identify top-X (say, 5) most used plugins and update them too (if I am correct that they depend on eslint validation code), and release it as a minor version change - some small share of users will be affected, they'll have to lock the version, we can then fix any reported issues.

To make any such change easier in the future it might be good to split validation code to the separate package, something like schema-utils in webpack, in which case eslint and plugins would depend on different versions of validation utils, and can be upgraded independently.

Or we could take a cautious route and release it as a major version, but then users will be carrying 2 ajv versions for longer (and they would still not be able to migrate eslint until plugins migrate). I think npm have finally fixed that old issue when two different sub-dependency versions were breaking installation, so it's not such a big deal too.

Your call.

File: lib/rules/no-restricted-modules.js
Comment (mdjermanovic) [reply lib/rules/no-restricted-modules.js:68]: I'll check more in detail. Some seem to be just useless, some look as if they should be 1-tuples.

If we decide to merge some fixes for a minor version, those fixes would be probably only removing useless `additionalItems`. Per our semver policy, a schema bug fix in a minor version shouldn't break end users, even if their config has de facto invalid settings.

**Comment by mdjermanovic** [comment]
> No, it allows ANY data type, "property"/"items" only applies any restriction IF it is "object"/"array" - this is how JSON Schema specifies it. So if the data were an object/array it would validate it as expected, otherwise it would just allow it (incorrectly). It is a common mistake, that's why I made strict mode default.

Thanks! I think I understand now. `"type"` in schema validates type of the instance. `"properties"` validate properties of the instance if the instance is an object (regardless of `"type"` in the schema), otherwise do nothing.

> * if plugins rely on eslint code validating their config schemas (am I correct - is this how it is?)

Plugin rules usually define `schema` property and thus rely on eslint validating user's configuraton for the rule. It's an optional property - rule doesn't have to define `schema` and it will be still getting user's configuration (without any validations by eslint). A plugin could implement its own validation, though I don't think there are any plugins doing this.

The schema itself isn't validated "in production". Only `RuleTester` (a built-in tool for the development of rules) validates rule schemas.

**Comment by epoberezkin** [comment]
> Plugin rules usually define schema property and thus rely on eslint validating user's configuraton for the rule. It's an optional property - rule doesn't have to define schema and it will be still getting user's configuration (without any validations by eslint). A plugin could implement its own validation, though I don't think there are any plugins doing this.

That’s what I thought.

So it means that the plugins that have “id” or “exclusiveMaximum/Minimum” (it was boolean in draft-04, number now) would throw exception. I think both should be quite rare.

The possible solution could be to add some code that would migrate plug-in schemas on the fly by replacing these two things (or we could just plug in json-schema-migrate) - what do you think?

> "type" in schema validates type of the instance. "properties" validate properties of the instance if the instance is an object (regardless of "type" in the schema), otherwise do nothing.

Exactly. So if you want to do minor version upgrade, we could avoid adding “type”, so that users validation don’t fail even if they have incorrect config.

**Comment by mdjermanovic** [comment]
> > > strict mode is on by default now
> 
> > Should we then turn it off explicitly, to retain the previous behavior (and thus maybe avoid breaking plugins)?
> 
> We could, but:
> 
> * there may be other schema mistakes that would remain unnoticed

There are two scenarios where schemas can be validated:

* runtime (end users) - indirectly via `compile()`.
* development (of ESLint core and plugins) - `RuleTester` runs `validateSchema()` and `compile()` with bit stricter options.

For end users, we already have [validateSchema: false](https://github.com/eslint/eslint/blob/06b58096975935ec016d96dd5f333f059c270f26/lib/shared/ajv.js#L22) (which disables the most severe errors?). I believe the intention was always to _not_ validate schemas in runtime. Enabling additional validations now will break end users that have valid configurations, so they'll have to wait for plugins to fix rule schemas and release new versions, although those plugins already work perfectly well for them.

For example, linting of eslint/eslint repo itself as an end-user would be broken until `eslint-plugin-eslint-plugin` and `eslint-plugin-node` fix schemas in some rules (I tried `npm run lint` on this branch, but also applied your changes in [eslint/eslintrc](https://github.com/eslint/eslintrc); we are extracting some functionality to that package, so we're currently using Ajv from two places).

I think we should turn off all validations of schemas in runtime, even for a major ESLint version, because it looks like it could cause a big breakage without clear benefits.

As for the development, it certainly makes sense to consider enabling additional validations in `RuleTester` in a major version. This also needs some discussion, because fixing schemas can also break end users of plugins, though only those with possibly invalid configurations.

(I'm aware that the incompatibility of schema versions is probably a bigger problem, just wanted to figure out how to handle validation changes first)

**Comment by mdjermanovic** [comment]
Also, it seems that `compile()` is more strict in Ajv 7.

For example, an invalid `type` value used to be ignored in Ajv 6, but it throws in Ajv 7:

```js
const Ajv = require("ajv").default;

const ajv = new Ajv({ 
  validateSchema: false,
  strict: false,
  strictTypes: false,
});

const schema = {
  type: "arrayy", // typo
};

ajv.compile(schema); // Error: type must be JSONType or JSONType[]: arrayy
```

**Comment by epoberezkin** [comment]
Yes, in general the approach was to be stricter by default (with many restrictions flipping from opt-in to opt-out), and treating possible typos as errors - it's been the source of many bugs. The logic is that JSON Schema is a language with constrained vocabulary and for the absolute majority of users the expected behaviour is to fail on unknown words (rather than ignore them as the spec suggests) - same as programming languages behave.

This blog post covers the scope of changes and the motivation: https://www.poberezkin.com/posts/2020-11-14-ajv-time-to-migrate-to-version-7.html

**Comment by nzakas** [comment]
This seem complicated enough that an RFC would seem to be in order to work out the details. Thoughts?

**Comment by mdjermanovic** [comment]
> I think npm have finally fixed that old issue when two different sub-dependency versions were breaking installation, so it's not such a big deal too.

There was an issue with `ajv-keywords` not getting the correct version of `ajv` as its peer dependency: https://github.com/eslint/eslint/pull/10022#issuecomment-368601457. 

I'm not sure if this has been fixed in npm < 7, but there's probably nothing we can do about it. Hopefully, it won't be happening since `table` doesn't have `ajv-keywords` as a dependency as of v5.

**Comment by epoberezkin** [comment]
I’ve made the PR for table to upgrade to Ajv v7 - it doesn’t have to be a major version change there.

**Comment by epoberezkin** [comment]
Happy New year all!

Ajv v7 is the main version since mid-December; no complaints about version conflicts so far. There's been a few minor bugs reported and fixed, mostly around new standalone code generation, - all good otherwise.

Table was just released yesterday with ajv v7 in the same major version - thanks @gajus.

Not sure how it is best to move it forward  - how would you decide between the options here: https://github.com/eslint/eslint/pull/13911#issuecomment-739720730 ?

There will be another major version release in the next 3-4 months, but it should be "mostly" backwards compatible.

File: package.json
Comment (brodybits) [inline package.json:51]:
```suggestion
    "ajv": "^7.0.3",
```

**Comment by nzakas** [comment]
@epoberezkin do you have a timeline for updating this to v8.0? We are starting to work on ESLint v8.0.0 and we'd like to get this in.

**Comment by epoberezkin** [comment]
@nzakas @mdjermanovic it's ready

Final changes:
- upgrade ajv to v8.5.0
- add ajv-draft-04 to keep JSON Schema draft-04 support
- add strict mode warning logs (excluding strictTuples warning)
- remove no-op "additionalItems" from schemas
- add missing "implied" types (that are not implied with JSON Schema)
- change property dataPath in ajv errors to instancePath
- update tests to for changed validation error messages

File: lib/shared/ajv.js
Comment (aladdin-add) [inline lib/shared/ajv.js:11]: the ajv v8 was listed in dependencies, but is it required in somewhere?

Comment (epoberezkin) [reply lib/shared/ajv.js:11]: It’s a peer dependency so that updating it doesn’t require ajv-draft-04 release

**Review by aladdin-add — APPROVED** [review]
LGTM, thanks!

**Comment by epoberezkin** [comment]
@aladdin-add thanks - when is it planned for release?

**Comment by aladdin-add** [comment]
friendly ping @mdjermanovic , has your concern been addressed?

File: lib/rule-tester/rule-tester.js
Comment (mdjermanovic) [inline lib/rule-tester/rule-tester.js:53]: This is a breaking change for plugins, because their tests will fail for all rules with schemas that don't pass [strictSchema validations](https://ajv.js.org/strict-mode.html#prohibit-ignored-keywords).

(enabling `strictDefaults` in `RuleTester` was also treated as a breaking change: https://github.com/eslint/eslint/issues/11473)

File: lib/shared/ajv.js
Comment (mdjermanovic) [inline lib/shared/ajv.js:22]: I think it doesn't make sense for end users to see logs about problems in schema definitions, as they have no control over it.

**Comment by mdjermanovic** [comment]
This is also a breaking change for end users. The new version of `Ajv#compile()` throws errors on some mistakes in the provided schema (e.g. https://github.com/eslint/eslint/pull/13911#issuecomment-741926204) that used to be ignored by Ajv 6.

We could add try-catch around `ajv.compile()` to mitigate this.

**Comment by mdjermanovic** [comment]
`lib/shared/ajv.js` is currently used by `RuleTester` only. Once we figure out what changes should be made, the same changes should be implemented in [@eslint/eslintrc](https://github.com/eslint/eslintrc) (that's where Ajv is used to validate user configurations).

**Comment by mdjermanovic** [comment]
I have concerns about `"ajv": "^8.5.0"` being declared as a [peer dependency](https://github.com/ajv-validator/ajv-draft-04/blob/2c3f50e9fbaf9e4b2bd7a86d84dc0575a5687a60/package.json#L40-L42) of `ajv-draft-04`, because `ajv-draft-04` [requires Ajv](https://github.com/ajv-validator/ajv-draft-04/blob/2c3f50e9fbaf9e4b2bd7a86d84dc0575a5687a60/src/index.ts#L2).

We can declare `"ajv": "^8.5.0"` as a dependency of ESLint, but will that guarantee that `require("ajv")` from `ajv-draft-04` will be getting the right version of Ajv when ESLint is installed as a dependency/dev-dependency of a project?

For example, I made a [test project](https://github.com/mdjermanovic/use-ajv-draft-04) with the following dependencies (it's using ESLint from this branch):

```js
"dependencies": {
    "ajv": "^6.10.0",
    "eslint": "github:epoberezkin/eslint#epoberezkin/ajv-v7"
  }
```

[Running ESLint](https://github.com/mdjermanovic/use-ajv-draft-04/blob/main/index.js) failes, with this error:

```text
Error: Cannot find module 'ajv/dist/core'
Require stack:
- /home/runner/work/use-ajv-draft-04/use-ajv-draft-04/node_modules/ajv-draft-04/dist/index.js
- /home/runner/work/use-ajv-draft-04/use-ajv-draft-04/node_modules/eslint/lib/shared/ajv.js
- /home/runner/work/use-ajv-draft-04/use-ajv-draft-04/node_modules/eslint/lib/shared/config-validator.js
- /home/runner/work/use-ajv-draft-04/use-ajv-draft-04/node_modules/eslint/lib/rule-tester/rule-tester.js
- /home/runner/work/use-ajv-draft-04/use-ajv-draft-04/node_modules/eslint/lib/rule-tester/index.js
- /home/runner/work/use-ajv-draft-04/use-ajv-draft-04/node_modules/eslint/lib/api.js
- /home/runner/work/use-ajv-draft-04/use-ajv-draft-04/index.js
```

because `npm install` creates the following structure:

```text
node_modules
├── ajv@6.12.6    
├── ajv-draft-04
└── eslint
    └── node_modules
        └── ajv@8.5.0
```

so `require("ajv/dist/core")` from `ajv-draft-04` gets resolved to Ajv v6.

Since `ajv-draft-04` directly `require`-s `ajv`, shouldn't `ajv` be declared as dependency of `ajv-draft-04`? (instead of _peer_ dependency)

**Comment by aladdin-add** [comment]
just moved it back to v8.0.0 project, as it seems a breaking change. :)

**Comment by epoberezkin** [comment]
> Since ajv-draft-04 directly require-s ajv, shouldn't ajv be declared as dependency of ajv-draft-04? (instead of peer dependency)

I will see how to change it… Possibly in this case it should be a direct dependency indeed.

**Comment by mdjermanovic** [comment]
> This is also a breaking change for end users. The new version of `Ajv#compile()` throws errors on some mistakes in the provided schema (e.g. [#13911 (comment)](https://github.com/eslint/eslint/pull/13911#issuecomment-741926204)) that used to be ignored by Ajv 6.
> 
> We could add try-catch around `ajv.compile()` to mitigate this.

Actually, try/catch may not be a good idea. It would be okay to just skip the validation, but it will also skip setting default values, so the rule might silently change behavior.

File: lib/shared/ajv.js
Comment (epoberezkin) [reply lib/shared/ajv.js:22]: let's remove it then :)

**Comment by epoberezkin** [comment]
> The new version of Ajv#compile() throws errors on some mistakes in the provided schema (e.g. #13911 (comment)) that used to be ignored by Ajv 6.

@mdjermanovic The example in the comment with `type: "arrayy"` actually throws exception in v6 as well, unless you disable meta-schema validation (which is enabled in eslent).

The only difference I can think of is the change from strictDefaults to strictKeywords.

If it makes it simpler (e.g. could be released as a non-breaking change), I could add this option (strictDefaults) back? Let me know.

Looking at ajv-draft-04 anyway.

**Comment by mdjermanovic** [comment]
> @mdjermanovic The example in the comment with `type: "arrayy"` actually throws exception in v6 as well, unless you disable meta-schema validation (which is enabled in eslent).

It's enabled in `RuleTester`, which is a tool for developers to test the rules. Rules with `type: "arrayy"` don't throw exceptions when they are actually used to lint code. We use one [setting for user config validations](https://github.com/eslint/eslint/blob/55c1a503404744a2d88ddd553fce6498acea2f99/lib/shared/ajv.js#L20-L25), and [another (more strict) in `RuleTester`](https://github.com/eslint/eslint/blob/55c1a503404744a2d88ddd553fce6498acea2f99/lib/rule-tester/rule-tester.js#L53) (+ [`ajv.validateSchema()` in `RuleTester`](https://github.com/eslint/eslint/blob/55c1a503404744a2d88ddd553fce6498acea2f99/lib/rule-tester/rule-tester.js#L532)). 

That said, we can assume that most plugins use `RuleTester` to test the rules, so the possibility that rules with ``type: "arrayy"`` were actually published should be very small.

**Comment by mdjermanovic** [comment]
> just moved it back to v8.0.0 project, as it seems a breaking change. :)

We're adding only issues to the project, to avoid duplicates. I removed this PR from the project now (the related issue https://github.com/eslint/eslint/issues/13888 is already there).

File: lib/shared/ajv.js
Comment (mdjermanovic) [inline lib/shared/ajv.js:23]:
```suggestion
        strict: false,
```

If we set `strict: false`, then we don't need to explicitly set `strictTuples: false`?

**Review by Gstar123456 — APPROVED** [review]

File: lib/shared/ajv.js
Comment (epoberezkin) [reply lib/shared/ajv.js:23]: Correct

File: lib/rules/no-restricted-modules.js
Comment (mdjermanovic) [reply lib/rules/no-restricted-modules.js:68]: I checked all removed `additionalItems: false`, and this is the only one that was supposed to limit the number of elements in an array. Others were just useless.

Given that this rule is deprecated, it's fine to remove the ignored `additionalItems: false` and not fix the schema.

**Comment by mdjermanovic** [comment]
> > Since ajv-draft-04 directly require-s ajv, shouldn't ajv be declared as dependency of ajv-draft-04? (instead of peer dependency)
> 
> I will see how to change it… Possibly in this case it should be a direct dependency indeed.

@epoberezkin any news on this? I think that's the only major issue. Aside from that, this looks good assuming we'll set `strict: false`.

**Comment by mdjermanovic** [comment]
I noticed that `npm test` is much slower on this branch, and located the cause in  RuleTester.

[`ajv.compile`](https://github.com/epoberezkin/eslint/blob/26574d8db6ec434df6b8ec0d525353f636bba126/lib/rule-tester/rule-tester.js#L551) takes ~ 3 minutes on this branch, vs only ~ 2 seconds on master.

I thought Ajv 8 might be slower due to the new validations, and made some benchmarks but they showed that Ajv 8 is actually _much_ faster.

The problem is probably here:

https://github.com/eslint/eslint/blob/278813a6e759f6b5512ac64c7530c9c51732e692/lib/rule-tester/rule-tester.js#L557

We're validating the rule's schema for each test case (that seems unnecessary, but may take some time to analyze and fix). If the rule uses ESLint's shorthand array schema, then `getRuleOptionsSchema` creates a new instance for each test case.

Ajv 6 used to cache compiled schemas by content, but Ajv 8 caches by reference?

If that's the case, we can move `const schema = getRuleOptionsSchema(rule);` to the upper scope. I tried that, and it reduces `npm test` time by almost half.

**Comment by mdjermanovic** [comment]
@epoberezkin I'm closing this PR as described in https://github.com/eslint/eslint/issues/13888#issuecomment-872591875, Thanks for the PR, and thanks for all the work on Ajv!

**Comment by epoberezkin** [comment]
@mdjermanovic sorry, missed these comments...

> Ajv 6 used to cache compiled schemas by content, but Ajv 8 caches by reference?

That is correct. There is a [reasonable argument](https://github.com/ajv-validator/ajv/issues/1098) for that, but it turns out to be a problem for quite a few users - e.g. https://github.com/ajv-validator/ajv/issues/1413

It indeed highlights incorrect cases when the schema is compiled multiple times, so I don't think it should be reversed...

> In today's TSC meeting, we have decided to stay on Ajv 6.

Are there notes? It would be great to formulate the changes needed that would make transition to the next major version easier - v6 is stable indeed, but there will be other dependencies going out of date, so at some point the migration would have to happen...

**Comment by nzakas** [comment]
See https://github.com/eslint/tsc-meetings for notes and transcripts of all TSC meetings.

**Comment by epoberezkin** [comment]
@nzakas - thank you - I will update the problem with peer dependency, and let's review for eslint v9. By then the main reason for migration could be security warnings in v6 dev dependencies (that of course can be addressed by additional v6 release:).

Users do migrate to v8 - I estimate that about 20% of users migrated (based on the download count of ajv-formats that's only needed for v8, accounting for the fact that some know dependencies do not need ajv-formats).

**Pull request closed without merging**
